### PR TITLE
ltex-ls 15.2.0 (new formula)

### PR DESCRIPTION
**Pull request by gzagatti**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by gzagatti** [comment]
CI raises the following error:

```sh
  Error:  Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:3.0.0:run (add-ltex-cli) on project ltexls: An Ant BuildException has occured: Execute failed: java.io.IOException: Cannot run program "python" (in directory "/tmp/ltex-ls-20220225-20481-8duwhb/ltex-ls-15.2.0"): error=2, No such file or directory
  Error:  around Ant part ...<exec executable="python">... @ 4:31 in /tmp/ltex-ls-20220225-20481-8duwhb/ltex-ls-15.2.0/target/antrun/build-main.xml
  Error:  -> [Help 1]
```

However, I'm not able to reproduce it in my local machine as the formula builds with no problem.

File: Formula/ltex-ls.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  sha256 "59209730cb9cda57756a5d52c6af459f026ca72c63488dee3cfd232e4cfbf70a"
+  license "MPL-2.0"
+  head "https://github.com/valentjn/ltex-ls.git", branch: "develop"
+
+  depends_on "maven" => :build
+  depends_on "python@3.9" => :build
+  depends_on "openjdk@11"
+
+  def install
+    ENV["JAVA_HOME"] = Formula["openjdk@11"].opt_prefix
```
Comment (branchvincent) [inline Formula/ltex-ls.rb:14]: it seems this relies on a `python` executable being available: https://github.com/valentjn/ltex-ls/blob/4954220ebd34df6c032ca646fd9bab5bb4d8231e/pom.xml#L376

we can add one to `PATH`:
```suggestion
    ENV.prepend_path "PATH", Formula["python@3.9"].opt_libexec/"bin"
    ENV["JAVA_HOME"] = Formula["openjdk@11"].opt_prefix
```

File: Formula/ltex-ls.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
+class LtexLs < Formula
+  desc "LSP for LanguageTool with support for Latex, Markdown and Others"
+  homepage "https://valentjn.github.io/ltex/"
+  url "https://github.com/valentjn/ltex-ls/archive/refs/tags/15.2.0.tar.gz"
+  sha256 "59209730cb9cda57756a5d52c6af459f026ca72c63488dee3cfd232e4cfbf70a"
+  license "MPL-2.0"
+  head "https://github.com/valentjn/ltex-ls.git", branch: "develop"
+
+  depends_on "maven" => :build
+  depends_on "python@3.9" => :build
```
Comment (branchvincent) [inline Formula/ltex-ls.rb:10]: can this use `python@3.10`?
```suggestion
  depends_on "python@3.10" => :build
```

**Comment by gzagatti** [comment]
The problem with the new build is that the command

```py
system Formula["python@3.10"].bin/"python3", "-u", "tools/createCompletionLists.py"
```

Attempt to write completion lists in a temporary folder which is then unwrittable in CI. The source code is [here](https://github.com/valentjn/ltex-ls/blob/4954220ebd34df6c032ca646fd9bab5bb4d8231e/tools/createCompletionLists.py#L93)

```sh
Invalid argument: Path is not writable: /var/folders/07/mmpkd6_x7fvbpl_msv10c7fh0000gp/T
```

Since the completion lists fail to write, the maven unit test that requires it then fails. I'm not sure if the completion lists are required for the application to work or it's only temporarily created for the unit tests.

In any case, I am pushing a build that does not execute the tests. Let's see how it does.

**Comment by gzagatti** [comment]
It turns out the completion list is necessary, but I figure out that by changing the environment variable `TMPDIR` in the build, I can get the completion list script to produce temporary directories in the build directory.

**Comment by gzagatti** [comment]
The problem in CI seems to be unrelated with the build process:

```sh
  Error: ltex-ls: no bottle available!
  You can try to install from source with:
    brew install --build-from-source ltex-ls
  Please note building from source is unsupported. You will encounter build
  failures with some formulae. If you experience any issues please create pull
  requests instead of asking for help on Homebrew's GitHub, Twitter or any other
  official channels.
```

I'm not sure how to overcome this. Any help would be appreciated.

**Comment by gzagatti** [comment]
The two CI processes fail because there is no bottle available. How can I create a bottle @branchvincent?

**Comment by branchvincent** [comment]
> there is no bottle available

hmm, this is a confusing error but the reason is that `openjdk@11` isn't bottled on Monterey. If it's possible, we can try building with `openjdk`

**Comment by gzagatti** [comment]
@branchvincent  thanks for your comment. It was super helpful. I've made the modifications and the formula can build with `openjdk` without a problem. I hope that it now passes all CI test.

**Comment by SMillerDev** [comment]
```
  Invalid argument: Path is not writable: /var/folders/sd/fmj8vmcx7f32h_wz5tn7dy4c0000gp/T
```

It's trying to write outside of the install prefix

**Comment by gzagatti** [comment]
I forgot to fix `JAVA_HOME` when I updated `openjdk`. I’m glad that fixing it solved all the issues.

**Comment by gzagatti** [comment]
I'm just wondering if there is any expectation for review of this PR?

File: Formula/ltex-ls.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    ENV["TMPDIR"] = buildpath
+    system Formula["python@3.10"].bin/"python3", "-u", "tools/createCompletionLists.py"
+    system "mvn", "-B", "-e", "-DskipTests", "package"
+    (buildpath/"build").mkpath
+    system "tar", "xzf", "target/ltex-ls-#{version}.tar.gz", "-C", "build"
+    # remove Windows files
+    rm Dir["build/ltex-ls-#{version}/bin/*.bat"]
+    bin.install Dir["build/ltex-ls-#{version}/bin/*"]
+    libexec.install Dir["build/ltex-ls-#{version}/*"]
+    bin.env_script_all_files libexec/"bin", Language::Java.overridable_java_home_env
```
Comment (SMillerDev) [inline Formula/ltex-ls.rb:25]: Let's use some whitespace to separate steps here. Additionally there is no need to specify the python path if it's in path anyway.
```suggestion
    ENV.prepend_path "PATH", Formula["python@3.10"].opt_libexec/"bin"
    ENV["JAVA_HOME"] = Formula["openjdk"].opt_prefix
    ENV["TMPDIR"] = buildpath

    system "python3", "-u", "tools/createCompletionLists.py"

    system "mvn", "-B", "-e", "-DskipTests", "package"

    mkdir "build" do
      system "tar", "xzf", "../target/ltex-ls-#{version}.tar.gz", "-C", "."

      # remove Windows files
      rm Dir["ltex-ls-#{version}/bin/*.bat"]
      bin.install Dir["ltex-ls-#{version}/bin/*"]
      libexec.install Dir["ltex-ls-#{version}/*"]
    end

    bin.env_script_all_files libexec/"bin", Language::Java.overridable_java_home_env
```

File: Formula/ltex-ls.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    EOS
+
+    (testpath/"expected").write <<~EOS
+      #{testpath}/test:1:5: info: The pronoun 'She' is usually used with a third-person or a past tense verb. [HE_VERB_AGR]
+      She say wrong.
+          Use 'says'
+          Use 'said'
+    EOS
+
+    got = `#{bin}/ltex-cli "#{testpath}/test"`
```
Comment (SMillerDev) [inline Formula/ltex-ls.rb:40]:
```suggestion
    got = shell_output("#{bin}/ltex-cli '#{testpath}/test'")
```

**Comment by gzagatti** [comment]
Thanks for revision! I closed the issue by mistake  (the page refreshed with the mouse right on top of the close button :( ) and then re-opened.

File: Formula/ltex-ls.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    EOS
+
+    (testpath/"expected").write <<~EOS
+      #{testpath}/test:1:5: info: The pronoun 'She' is usually used with a third-person or a past tense verb. [HE_VERB_AGR]
+      She say wrong.
+          Use 'says'
+          Use 'said'
+    EOS
+
+    got = shell_output("#{bin}/ltex-cli '#{testpath}/test'")
```
Comment (SMillerDev) [inline Formula/ltex-ls.rb:46]:
```suggestion
    got = shell_output("#{bin}/ltex-cli '#{testpath}/test'", 3)
```

**Review by SMillerDev — APPROVED** [review]
Thanks @gzagatti  ! Without contributions like yours it'd be impossible to keep homebrew going with the high standards that users have come to expect from the project. You can feel good knowing that you've made the world a tiny bit better for homebrew users around the world! 👍 🎉

**Pull request closed without merging**